design/open/lakefs-on-kv-testing-plan.md

```diff
@@ -0,0 +1,128 @@
+
+
+### ```pkg/actions```
+* DB is used to store actions runs and hook runs with results. Db is read by the actions service to handle requests from the various clients. For each run there is a new entry in `actions_run` table and an entry for each hook in `actions_run_hooks` table.
+Updates are done for commitID, as part of `post` hooks
+* Currently unit tests cover hooks run and verify the results in DB using the service reading functions. There are also error testings (hook failures)
+* 67.6% coverage
+* What is missing:
+  * DB Failure during hook action (TBD - what is the expected behavior? Is that interesting to test?)
+  * Performance under load - not sure a load on actions/hooks is relevant
+  * Concurrent actions/hooks execution, which derives concurrent DB accesses. This gets more interesting if load is involved (in case load is relevant)
+
+### ```pkg/auth```
+* DB is used to store installation metadata, users and policies
+* Only 6.8% coverage
+* Some benchmarks exists that covers effectivePolicies (multi-table JOIN)
+* What is missing:
+  * Need to increase code coverage
+  * Comparison benchmark to toggle between Tables and KV and verify there is no degradation
+  * Migration tests:
+    * verify that data is migrated correctly
+    * Authentication transactions in during migration (data that was written to tables is read from KV)
+
+### ```pkg/diagnostics```
+* Only read access
+* No test coverage. Need to verify all is working the same after migration
+
+### ```pkg/graveler/ref```
+* Branch locking (read)
+* Repos, branches, commits and tags - read/write
+* 95.5% testing coverage
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### ```pkg/graveler/retention```
+* No direct DB access. Accesses are done using ```pkg/graveler/ref```, which should make the DB transition seamless
+* 42.9% coverage in unit testing
+* Missing performance tests that might help detect degradation
+
+### ```pkg/graveler/staging```
+* 89.1% coverage
+* Covers all staged KV functionality
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### Res-Dump/Restore
+Currently this functionality is not covered directly, but since it relies on '''pkg/graveler/ref''' for DB access, the DB migration should be seamless
+It can be leveraged, however, to extend the cover of '''pkg/graveler/ref''' and for performance, as it is quite exhaustive (traverses all branches and commits and tags, per repo)
+
+## Execution Plan
+
+### KVM1
+* Data level migration tests infrastructure
+  * Migrate data from Table to KV, dumps both and compares
+  * Implement dumpers for `gateway_multiprts`
+  * Implement tests for '''pkg/multiparts'''
+* Infrastructure for running migration during a system test execution
+  * System test to run migration during multiparts upload
+  * Currently there is a single simple multiparts system test (single file, 7 parts) - this is also an opportunity to expand that
```

Comment:
What would you expand here and why?

Reply:
The current test verifies a very specific simple good path.
I thought of adding, at least, a test that uploads multiple files with multiple parts each (e.g. 100 files with 1000 parts each). This will insert a concurrency aspect to the tests, which is currently lacks
I also thought of shuffeling the parts order, rather than upload them ordered sequentially, just for the fun of it
Another aspect, I miss in the current test, is errors and retries of uploads, so this, too can be added.
This is orthogonal to our KV needs and can (and IMHO should) be added' regardless. It will, however, help with achieving higher confidence in our KV support for the multiplrt pkg
WDYT?

Reply:
I don't think it's a must for the current milestone. Maybe leave it as a stretch goal?

Reply:
To the expansion.

Reply:
How about keeping it as a stretch goal?